### ISSUE TITLE
keys() to pick() typing problem

### ISSUE THREAD

**Issue by mirimCZ** [ramda/types#76]
tldr: passing result of `keys()` into `pick()` reports an error, it can be solved by doing some quite dirty casting, better fix could be by updating the typing in `pick` (see Fix block bellow), but I´m not sure of the intentions of current implementation - thus this issue to discuss it.

```typescript
/*
 * The usecase is I have an object that gets partially modified on FE
 * changes can be "flat" (on simple props like value), or also "nested" (eg on obj.body)
 */
interface Alpha {
  _id: string;
  name: string;
  value: string;
  description: string;
  obj: {
    head: string;
    body: string;
  }
}

interface AlphaPatch extends PartialDeep<Alpha> {
  // to make ID required, we need it for db filter
  _id: string;
}

function AlphaTest() {
  // in real app patch comes from the FE in Request body
  const alphaPatch: AlphaPatch = {
    _id: 'id',
    name: 'patched name',
    obj: {
      body: 'patched body',
    }
  };

  // in real app complete object is loaded from db, _id from patch is used for identification to load it
  const alpha: Alpha = {
    _id: 'id',
    name: 'name',
    value: 'value',
    description: 'description',
    obj: {
      head: 'obj.head',
      body: 'obj.body',
    },
  };

  /*
   * now because I dont want to just deep merge whole alpha with patch but I still need deep values in my patch
   * I'm trying to pick the keys, but pick reports an error:
   * Argument of type '("name" | "value" | "description" | "obj")[]' is not assignable to parameter of type 'readonly [keyof Alpha, ...(keyof Alpha)[]]'.
   * Source provides no match for required element at position 0 in target.ts(2345)
   */
  const { _id, ...patch } = alphaPatch;
  const keysToUpdate = keys(patch);
  const currentAlpha = pick(keysToUpdate, alpha);
  const patchWithAlphaDeep = mergeDeepRight(currentAlpha, patch);
}
```
Fix:
```typescript
// this is current typings for pick
export function pick<U, const Names extends readonly [keyof U, ...(keyof U)[]]>(names: Names, obj: U): string extends keyof U ? Record<string, U[keyof U]> : Pick<U, ElementOf<Names>>;

// this will fix it, look at Names;
declare module 'ramda' {
  export function pick<U, const Names extends readonly (keyof U)[]>(names: Names, obj: U): string extends keyof U ? Record<string, U[keyof U]> : Pick<U, ElementOf<Names>>;
}
```

I'm not sure whether this typing on pick is intentional or not / what kind of purpose this typings serves, but anyway it doesn´t seems right that it´s impossible to pass result of keys() into pick(). I did some experiments with trying to recasting it, I made it "work" but the type wasn't reflecting reality `keys(patch) as [keyof PartialDeep<Alpha>];` - that´s not right, there can be more than just 1 item in the array. Following cast will be correct, but it just feels dirty `keys(patch) as [keyof PartialDeep<Alpha>, ...(keyof PartialDeep<Alpha>)[]];`.

So my question is - is there a reasoning for having the typing for pick this way?

I found this change three weeks ago: https://github.com/ramda/types/pull/52/files @Harris-Miller

**Comment by Harris-Miller** [comment]
The `[keyof U, ...(keyof U)[]]` part was meant to prevent usage of `pick` with an empty array. However, typing `pick` that way does prevent when you have an array of known keys but unknown length. You're use-case _should_ be supported, but it's not, and we need to fix that

I don't have a lot of free time ATM to do this work myself right now, PRs welcome!

**Comment by Harris-Miller** [comment]
https://github.com/ramda/types/pull/78
